### PR TITLE
builtin: support uid/gid config for binfs app

### PR DESCRIPTION
**Pull request by fxysunshine**
Implement I_SUID/I_SGID feature for binfs in the POSIX compliant way. If set-user-ID bit is set in the file permissions, then the effective user ID of process shall be set to UID of the new process image file.

test case:
hello example emulates to set uid and file set-user-ID bit, and call geteuid and getegid API.
UID = 2000
GID = 3000
MODE = 06555

nsh> ls -l /bin/hello
-r-sr-sr-x    2000    3000       0 hello
nsh> hello
geteuid:2000
getegid:3000

## Testing
sim & pass CI

### CODE REVIEW

File: include/nuttx/binfmt/binfmt.h [hunk 1]
```diff
@@ -97,6 +97,7 @@ struct binary_s
 
   uint8_t priority;                    /* Task execution priority */
   size_t stacksize;                    /* Size of the stack in bytes (unallocated) */
+  uid_t proguid;                       /* User identity */
```
Comment (patacongo) [inline include/nuttx/binfmt/binfmt.h:100]: How does this proguid get set?

File: binfmt/binfmt_execmodule.c [hunk 2]
```diff
@@ -266,6 +266,10 @@ int exec_module(FAR struct binary_s *binp,
 
   pid = tcb->cmn.pid;
 
+#ifdef CONFIG_SCHED_USER_IDENTITY
+  tcb->cmn.group->tg_uid = binp->proguid;
```
Comment (patacongo) [inline binfmt/binfmt_execmodule.c:270]: Shouldn't the UID be inherited from the parent when the the builtin app is started?

File: binfmt/builtin.c [hunk 3]
```diff
@@ -106,6 +106,8 @@ static int builtin_loadbinary(FAR struct binary_s *binp,
   binp->entrypt   = builtin->main;
   binp->stacksize = builtin->stacksize;
   binp->priority  = builtin->priority;
+  binp->proguid   = builtin->proguid;
+
```
Comment (patacongo) [inline binfmt/builtin.c:110]: Same question:  Shouldn't the UID come from the UID of the parent?

File: sched/Kconfig [hunk 4]
```diff
@@ -622,6 +622,12 @@ config SCHED_THREAD_LOCAL
 		This option enables architecture-specific TLS support (__thread/thread_local keyword)
 		Note: Toolchain must be compiled with '--enable-tls' enabled
 
+config UID_DEFAULT
```
Comment (patacongo) [inline sched/Kconfig:625]: Where is CONFIG_UID_DEFAULT used?

File: include/nuttx/binfmt/binfmt.h (hunk 1 above)
Comment (fxysunshine) [reply include/nuttx/binfmt/binfmt.h:100]: The apps side commit need follow.
https://github.com/apache/nuttx-apps/pull/1691/

File: binfmt/binfmt_execmodule.c (hunk 2 above)
Comment (fxysunshine) [reply binfmt/binfmt_execmodule.c:270]: I think that is depended on the fork and setuid policy from app. But we can set a default UID when task is creating.

File: sched/Kconfig (hunk 4 above)
Comment (fxysunshine) [reply sched/Kconfig:625]: used by builtin apps side if not configured.
https://github.com/apache/nuttx-apps/pull/1691

Comment (patacongo) [reply sched/Kconfig:625]: > used by builtin apps side if not configured.

If it effects apps/ (which are optional), then it should not be in an OS Kconfig file.  It should be an an apps/ Kconfig file.

File: binfmt/binfmt_execmodule.c (hunk 2 above)
Comment (patacongo) [reply binfmt/binfmt_execmodule.c:270]: > I think that is depended on the fork and setuid policy from app. But we can set a default UID when task is creating.

Everything I have read states that the UID should be the UID of the parent.  setuid() should be able to change the UID.

In Unix systems there is an effective UID (euid) and a real UID (ruid).  Both are inherited from the parent.  The child can change is effective UID but not its real UID.

I certainly think we as a project need to agree on this policy before we commit to it.

File: sched/Kconfig (hunk 4 above)
Comment (cederom) [reply sched/Kconfig:625]: I think this idea comes from Android where each application runs on its own UID? Do I get the idea right?

What if we set `UID_DEFAULT` to `0` (`root`)?

**Comment by patacongo** [comment]
It would help to understand exactly what you are trying to accomplish to forcing some UID defined in a configuration.

NuttX supports standard /etc/passwd and /etc/group files for assigning UIDs and GIDs to users.  These are currently only used in NSH logins and do not for set UID and GID, but they should!  See configuration sim:nsh, boards/sim/sim/sim/include/nsh_romfsimg.h, and boards/sim/sim/sim/src/etc.

The behavior that I would expect is that the UID and GID defaut to zero (root) but are set at login to the values in /etc.  Tasks run (in memory or in the file system) would (usually) take the UID and GID of the parent. But could be changed with ST_NOSUID logic or setuid() and setgid().

**Comment by cederom** [comment]
Good question @patacongo I don't get it too :-) 

Could you please elaborate more on your goal @fxysunshine ?

In Unix if you want to provide different applications with a separate access levels (i.e. to filesystem), for instance when you run a service, you would want to create a dedicated user and group for that application (or group of applications), then launch that application by that user/group. Services may be launched by system init scripts and/or local user crontab. Not sure if all this functionality is already here in NuttX?

The problem you are facing right now is that UID/GID is not set even if you run the application by a given user?

Maybe you want something like `su` to run one application as another user?

Maybe you want `setuid`/`setgid` functionality in order to spawn process with `uid`/`gid` or a file owner?

**Comment by patacongo** [comment]
> The problem you are facing right now is that UID/GID is not set even if you run the application by a given user?

Currently, each task group does inherit the IDs of its parent.  But everything is zero (root) so it doesn't do much,  A lot of machinery is in place to do good Unix-style access control (at least in a KERNEL build).  But it is not being developed.  We really need a  published roadmap if where we are and where we want to go.  I worry about _ad hoc_ solutions that are not based on a long term, POSIX compliant roadmap.

**Comment by cederom** [comment]
> @patacongo: We really need a published roadmap if where we are and where we want to go. I worry about _ad hoc_ solutions that are not based on a long term, POSIX compliant roadmap.

I usually do a simple TODO list with points and subpoints and subpoints for eatch project in a simple wiki style MarkDown (example below). Then I know what needs to be done, who is working on it, how much time it cost, what are the reference pr, etc. This is a really simple file easy to edit. Things may pop up there as discovered by new (sub)tasks. And it is a good reference point / worklog in future. It makes vision, planning, implementation, and accounting easier :-)

Maybe this could be implemented for NuttX (either as part of the source code text document, or GitGub wiki, or project wiki)?

I know there are tools like Jira.. but the general roadmap should be easy to read and containted in one place easy to edit / update.

If there is a consensus on how to implement such roadmap I could help in creating / updating one just let me know :-)

# TODO / Roadmap
* Kernel.
  *  [ ] implement pid + gid assignment to process inherited by uid/git of the launching user.
      * [ ] commit.
      * [ ] testing.
      * [ ] implementation.
      *  [X] design discussion.
* Application.
  * [ ] implement `su`.
* Bootloader.
  * [X] fix some bootloader stuff.

**Comment by patacongo** [comment]
> TODO / Roadmap

This is a good discussion, but we should probably move it to a separate issue so that we don't overwhelm this topic.

File: binfmt/binfmt_execmodule.c (hunk 2 above)
Comment (fxysunshine) [reply binfmt/binfmt_execmodule.c:270]: > NuttX currently sets the user and group IDs of "normally" created tasks to the IDs of the parent: https://github.com/apache/nuttx/blob/master/sched/group/group_create.c#L91 I think an inconsistent policy for executable modules would be a bad idea.
> 
> When running executable files on a file system, the POSIX behavior differs depending on the state of ST_NOSUID mount flags on the volume. Are you trying to emulate setting the UID from the UID in the executable file permissions?

Thank you for explaining posix uid/gid in detail. I think file system that supports set-user-ID permission bits is what we want, and also it follows POSIX policy.

We need a program UID and the idea comes from Android. The UID maybe a holder, not real user UID and exist in /etc/passwd. So the different program can use getuid() from what they configured. The client sdk can know what it is UID from different exe programs, and so the sdk can isolate some user data or permission control for some policy. 

If ST_NOSUID is set on the mounted file system, we can set a UID that similar with linux chown. Maybe use geteuid() to get the UID what we need, and getuid() is alway the real UID of calling process.

Do you think that we do some work to make this happen in Nuttx, and also the solution is compatible with POSIX？

File: sched/Kconfig (hunk 4 above)
Comment (fxysunshine) [reply sched/Kconfig:625]: > I think this idea comes from Android where each application runs on its own UID? Do I get the idea right?

Yes, we use UID to isolate or share some user data, permission control in some complicated scenes.  But it is not being developed in Nuttx kernel.

**Comment by xiaoxiang781216** [comment]
@patacongo we want to implement I_SUID/I_SGID feature for binfs and romfs.
Why we need these features? Because the traditional security model like your description:

> It would help to understand exactly what you are trying to accomplish to forcing some UID defined in a configuration.
> 
> NuttX supports standard /etc/passwd and /etc/group files for assigning UIDs and GIDs to users. These are currently only used in NSH logins and do not for set UID and GID, but they should! See configuration sim:nsh, boards/sim/sim/sim/include/nsh_romfsimg.h, and boards/sim/sim/sim/src/etc.
> 
> The behavior that I would expect is that the UID and GID defaut to zero (root) but are set at login to the values in /etc. Tasks run (in memory or in the file system) would (usually) take the UID and GID of the parent. But could be changed with ST_NOSUID logic or setuid() and setgid().

isn't suitable for personal device(e.g. PC, phone and PAD). The old model assume that software is perfect, and the device is shared by many people, the security mode trusts the installed software but untrust the end user. So, it's good approach to assign each login user a different id and do the security check in various syscall. But for personal device, the user owns the device physically, it's make no sense to check the user activity, but program downloaded from internet which may contain malicious code need be monitored instead.

Here is how Android use uid/gid to protect the user from maliciour program: 
https://stackoverflow.com/questions/5708906/what-is-uid-on-android
Basically, Android assign program with the different uid/gid and monitor the program activity in many places.  I_SUID/I_SGID is the most portable and simplest way to achieve this goal. 

@cederom Yes, we want to do the similar thing on NuttX like Android on Linux in the POSIX compliant way.

**Comment by fxysunshine** [comment]
> @patacongo we want to implement I_SUID/I_SGID feature for binfs and romfs. Why we need these features? Because the traditional security model like your description:
> 
> > It would help to understand exactly what you are trying to accomplish to forcing some UID defined in a configuration.
> > NuttX supports standard /etc/passwd and /etc/group files for assigning UIDs and GIDs to users. These are currently only used in NSH logins and do not for set UID and GID, but they should! See configuration sim:nsh, boards/sim/sim/sim/include/nsh_romfsimg.h, and boards/sim/sim/sim/src/etc.
> > The behavior that I would expect is that the UID and GID defaut to zero (root) but are set at login to the values in /etc. Tasks run (in memory or in the file system) would (usually) take the UID and GID of the parent. But could be changed with ST_NOSUID logic or setuid() and setgid().
> 
> isn't suitable for personal device(e.g. PC, phone and PAD). The old model assume that software is perfect, and the device is shared by many people, the security mode trusts the installed software but untrust the end user. So, it's good approach to assign each login user a different id and do the security check in various syscall. But for personal device, the user owns the device physically, it's make no sense to check the user activity, but program downloaded from internet which may contain malicious code need be monitored instead.
> 
> Here is how Android use uid/gid to protect the user from maliciour program: https://stackoverflow.com/questions/5708906/what-is-uid-on-android Basically, Android assign program with the different uid/gid and monitor the program activity in many places. I_SUID/I_SGID is the most portable and simplest way to achieve this goal.
> 
> @cederom Yes, we want to do the similar thing on NuttX like Android on Linux in the POSIX compliant way.

@patacongo @cederom have done some rework  to follow POSIX spec. For builtin app, We add uid/gid/mode in compile time to builtin_s struct, builtin loader checks the file mode I_SUID/I_SGID and configure the euid/egid from file's uid/gid. 
If binfs is enabled, euid/egid/mode export to file system,  so we can show file owner and group id, for example "ls" command.
Could you nice to review?

File: sched/group/group_setegid.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  /* Get the currently executing thread's task group. */
+
+  rtcb   = this_task();
+  rgroup = rtcb->group;
+
+  /* Set the task group's group identity. */
+
+  DEBUGASSERT(rgroup != NULL);
+  rgroup->tg_egid = gid;
+  return OK;
```
Comment (patacongo) [inline sched/group/group_setegid.c:83]: Eventually, this needs to be fixed. And unprivileged task should not be permitted to change its effective GID.

**Review by patacongo — COMMENTED** [review]
I think this is an interesting change.  My only real concern is that the scope of the change is too limited.  It is too focused on your immediate need and does not address the general feature.

S_ISUID and S_ISGUID apply to executable files ran from all file systems included ELF models and NXFLAT modules.  This change restricts the support to builtin applications..

Exending support to other file systems should be a simple matter since they all have a common interface;  they do not have to be modified individually.

Of particular concern is binfs which is a special file system wrapper for builtin applications.  It would need some custom work in order to propagate the S_ISUID and S_ISGID information into the file mode.

File: libs/libc/builtin/lib_builtin_getgid.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  builtin = builtin_for_index(index);
+
+  if (builtin != NULL)
+    {
+      return builtin->gid;
+    }
+
+  /* Return group user identity 'root' with a gid value of 0. */
+
+  return 0;
```
Comment (patacongo) [inline libs/libc/builtin/lib_builtin_getgid.c:62]: How does the builtin GID and UID get set?  Is that a companion PR for nuttx-apps that does this?

Comment (fxysunshine) [reply libs/libc/builtin/lib_builtin_getgid.c:62]: YES,   https://github.com/apache/nuttx-apps/pull/1691

File: fs/binfs/fs_binfs.c [hunk 7]
```diff
@@ -470,21 +470,35 @@ static int binfs_stat(struct inode *mountpt,
...
       buf->st_mode = S_IFREG | S_IXOTH | S_IXGRP | S_IXUSR;
+
+#ifdef CONFIG_SCHED_USER_IDENTITY
+      buf->st_uid = builtin_getuid(index);
+      buf->st_gid = builtin_getgid(index);
+
+      mode = builtin_getmode(index);
+      buf->st_mode |= (mode & S_ISUID) ? S_ISUID : 0;
+      buf->st_mode |= (mode & S_ISGID) ? S_ISGID : 0;
+#endif
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply fs/binfs/fs_binfs.c:501]: Let's append a new patch to enable SUID/SGID for elf loader too

File: sched/group/group_seteuid.c [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  /* Get the currently executing thread's task group. */
+
+  rtcb   = this_task();
+  rgroup = rtcb->group;
+
+  /* Set the task group's group identity. */
+
+  DEBUGASSERT(rgroup != NULL);
+  rgroup->tg_euid = uid;
+  return OK;
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply sched/group/group_seteuid.c:85]: ditto

**Comment by xiaoxiang781216** [comment]
> Many builtin applications do not use binfs and hence would not use the logic that have implemented. Many builtin applications are executed from the NSH command line and use only logic under nuttx-apps/builtin
> 
> So the S_ISUID and S_ISGID logic would not work if the the very same application were executed from NSH without using binfs.

Yes, it's hard to support SUID/SGID without binfs because posix_spawn doesn't have parameters to pass the custom uid/gid.

File: binfmt/builtin.c [hunk 9]
```diff
@@ -106,6 +106,12 @@ static int builtin_loadbinary(FAR struct binary_s *binp,
   binp->entrypt   = builtin->main;
   binp->stacksize = builtin->stacksize;
   binp->priority  = builtin->priority;
+#ifdef CONFIG_SCHED_USER_IDENTITY
+  binp->uid       = builtin->uid;
+  binp->gid       = builtin->gid;
+  binp->mode      = builtin->mode;
+#endif
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply binfmt/builtin.c:113]: all binp is zero by kmm_zalloc

[the remark this answers is not in the document]
Comment (patacongo) [reply binfmt/builtin.c:113]: > all binp is zero by kmm_zalloc

Shouldn't the values come from the group structure?  If S_ISUID/GID is not set or not supported, then the child should inherit the UID/GID of the parent.

[the remark this answers is not in the document]
Comment (patacongo) [reply binfmt/builtin.c:113]: > I suppose this does not matter much right now since everything runs as root:root

When the user logs into NSH, the UID and GID should be available in /etc/password.  These should be used to set the session UID/GID sometime in the future.  Then inheritance of the UID/GID would be more important.

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply binfmt/builtin.c:113]: > > all binp is zero by kmm_zalloc
> 
> Shouldn't the values come from the group structure?

No, values in binp come from the binfmt provider(e.g. binfs from builtin table, elf come from file system).

> If S_ISUID/GID is not set or not supported, then the child should inherit the UID/GID of the parent.
> 

Yes, UID/GID in the new task tcb is set to parent's value always when we create task.

> I suppose this does not matter much right now since everything runs as root:root

code in the common binfmt change to binfs/elf provided one:
https://github.com/apache/nuttx/pull/8924/files#diff-2712024943f67847379bc555b8cd95ac99d41ca9e009fc746ce991527ea76809R283-R293

So, the sequence is:

1. The new task tcb get the initial uid/gid from parent task
2. binfs/elf set uid/gid in binp after getting info from builtin table for file system
3. binfmt change tcb's uid/gid with binp's values if S_UID/S_GID is set in mode field

File: sched/group/group_setegid.c (hunk 5 above)
Comment (xiaoxiang781216) [reply sched/group/group_setegid.c:83]: > > Yes, it's hard to support SUID/SGID without binfs because posix_spawn doesn't have parameters to pass the custom uid/gid.
> 
> They are available in the builtin registry data, that is how they are obtained in the binfs case.

it's simple toget this info in nsh, but how we pass this info to kernel is hard without modifying task_create or task_spawn.

**Comment by xiaoxiang781216** [comment]
> Many builtin applications do not use binfs and hence would not use the logic that have implemented. Many builtin applications are executed from the NSH command line and use only logic under nuttx-apps/builtin
> 
> So the S_ISUID and S_ISGID logic would not work if the the very same application were executed from NSH without using binfs.

Yes, if we want to make this feature work without binfs, we have to add mode/uid/gid to task_create or task_spawn.

File: sched/group/group_setegid.c (hunk 5 above)
Comment (xiaoxiang781216) [reply sched/group/group_setegid.c:83]: It's the independent but related job:

1. Assign the task with the correct uid/gid from user login or SUID/SGID
2. Do the permission check in various place

Item 2 need the work from item1, but item 2 need more work. We need do the huge work step by step, otherwise nothing will happen.

**Comment by fxysunshine** [comment]
binfmt: support euid of process set from the file system #10176  https://github.com/apache/nuttx/pull/10176
sched: implement effective uid and gid interfaces #10119  https://github.com/apache/nuttx/pull/10119.

File: fs/binfs/fs_binfs.c [hunk 10]
```diff
@@ -469,21 +469,35 @@ static int binfs_stat(struct inode *mountpt,
...
 
       /* It's a execute-only file name */
 
       buf->st_mode = S_IFREG | S_IXOTH | S_IXGRP | S_IXUSR;
+
+#ifdef CONFIG_SCHED_USER_IDENTITY
+      buf->st_uid = builtin_getuid(index);
+      buf->st_gid = builtin_getgid(index);
+
+      mode = builtin_getmode(index);
```
Comment (xiaoxiang781216) [inline fs/binfs/fs_binfs.c:497]: let's apply the suggestion:
https://github.com/apache/nuttx/pull/10203#discussion_r1292201964
https://github.com/apache/nuttx-apps/pull/1691/files#r1292692039

Comment (fxysunshine) [reply fs/binfs/fs_binfs.c]: we can not set the file type in the mode, so need append file type S_ISREG for binfs mode

**Review by xiaoxiang781216 — APPROVED** [review]
